_posts/2025-01-30-introducing-sealed-types.markdown

```diff
@@ -0,0 +1,103 @@
+
+# Using sealing for opaque types
+
+In CHERI systems, you can *seal* a CHERI capability (a pointer) and then it becomes an opaque token.
+CHERI capabilities contain an object type field that, for sealed capabilities, contains a value taken from the capability used to seal them.
+
+You can store it in memory, copy it around, but you can't use it.
+The only operation that you can do with a sealed capability is unseal it, and you can do that only with the correct capability that authorises unsealing *this specific kind of* sealed capability.
+In C, sealing might look something like this:
+
+```c
+char *ptr = strdup("hello world");
+char *sealed = cheri_seal(ptr, someKey);
+// Don't do this, it will trap:
+// sealed[0] = 'H';
+// This will not trap, but it will clear the tag:
+// sealed = sealed + 1;
+// Unsealling with the correct key gives the original
+char *unsealed = cheri_unseal(sealed, someKey);
+assert(cheri_is_equal_exact(ptr, unsealed));
+// Unsealling with the wrong key gives an untagged value
+char *untagged = cheri_unseal(sealed, wrongKey);
+assert(cheri_tag_get(untagged) == 0);
+```
+
+Sealed pointers are subject to the same temporal safety mechanisms as any other allocation on CHERIoT and so implementing safe opaque pointers with sealing is easy.
+You create a sealed allocation and simply the sealed capability out of your compartment.
+When you receive one, unseal it and if the unseal succeeds it is a valid value of the type that you expect.
+When you free the object, all pointers to it (sealed or unsealed) become untagged and so all outstanding handles are implicitly invalidated.
+
+In CHERIoT RTOS, we used `SObj*` as a placeholder for all sealed types.
+This meant that we got compile errors if you tried to use them in most invalid ways but it also meant that you couldn't differentiate between two sealed types, such as an allocator capability and one that authorises access to a message queue.
+This was annoying because some APIs needed to take two different sealed types as arguments.
+It was also somewhat annoying because sealing enforces type safety, but you needed to explicitly specify the type that you were trying to unseal an object as.
+
+# Sealed pointers in the type system
+
+[PR 109](https://github.com/CHERIoT-Platform/llvm-project/pull/109) in CHERIoT LLVM introduced `__sealed_capability` as a type qualifier.
+Attempting to dereference a sealed capability or modify its address is an error and will now fail at compile time, rather than run time.
+This includes modifications via the CHERI builtins.
+
+The seal builtin now returns a `__sealed_capability`-qualified pointer and the unseal builtin requires one.
+This means that, rather than making every sealed pointer a `SObj*`, we can make an allocator capability a `struct AllocatorCapabilityState *__sealed_capability)` and, when we unseal it, get back a `struct                    AllocatorCapabilityState *`.
```

Comment:
```suggestion
This means that, rather than making every sealed pointer a `SObj*`, we can make an allocator capability a `struct AllocatorCapabilityState *__sealed_capability)` and, when we unseal it, get back a `struct AllocatorCapabilityState *`.
```